File: lang/js/gdpjs/gdpjs_supt.c
/* vim: set ai sw=4 sts=4 ts=4 : */

// Support routines for Node.js Javascript programs accessing the
// GDP dynamic libraries, libgdp.xxx and libep.xxx as well as some
// GDP macros, and system (e.g., libc) routines.
//
// See libgdp_h.js for Node.js ffi "declarations" for these C routines
// to allow invoking them from JS.
//
// Alec Dara-Abrams
//  2014-10-28


#include "gdpjs_supt.h"
#include "gdp/gdp.h"
#include "ep/ep.h"
#include "ep/ep_stat.h"
// Seems to be a problem defining EP_STAT if ep/ep.h is not included before
// ep/ep_stat.h .  Check includes in ep_stat.h .


// DEBUG = 1 to include fprintf(stderr, ) debugging output; else = 0.
#define DEBUG 0


// EP_STAT_INT_EQUIV is needed for some casts below.
// We determine this by examning EP_STAT in ep/ep_stat.h .
// We assume these BIT counts exhaust EP_STAT.
#define EP_STAT_NBITS  ( _EP_STAT_SEVBITS + _EP_STAT_REGBITS + \
                         _EP_STAT_MODBITS + _EP_STAT_DETBITS )

#if   (EP_STAT_NBITS ==  32)
	#define EP_STAT_INT_EQUIV  uint32_t
#elif (EP_STAT_NBITS ==  64)
	#define EP_STAT_INT_EQUIV  uint64_t
#else
	#error "The type EP_STAT (in ep/ep_stat.h) must be 32 or 64 bits long"
#endif

// EP_STAT_32_64 is needed for some casts below
// Note, even though EP_STAT is all caps, it is a C typedef struct, not a
// cpp macro.  Yes, we may be making some assumptions here about padding, etc.
// but we only use this type for casts, and don't access internal structure.
typedef union
		{ EP_STAT            as_EP_STAT;
	      EP_STAT_INT_EQUIV  as_32_64_t;
        } EP_STAT_32_64;



// Some general libc-related functions

// Forces a flush() on stdout, on stderr, and on all open file descriptors.
// Node.js console.log(), console.error(), and process.stdxxx.write() are,
// evidently, buffered when writing to a file.  Output to a terminal seems
// to be unbuffered.
int fflush_stdout() { return ( fflush(stdout) ); }
int fflush_stderr() { return ( fflush(stderr) ); }
int fflush_all()    { return ( fflush( NULL ) ); }



// Some general libgdp/libep-related functions


// Returns the size in bytes of the libep error status code, EP_STAT .
size_t
sizeof_EP_STAT_in_bytes()
{
#if DEBUG
    fprintf( stderr,
	         "sizeof_EP_STAT_in_bytes: sizeof(EP_STAT) = %lu (decimal-%%lu)\n",
	         sizeof(EP_STAT)
		   );
#endif
	return (sizeof(EP_STAT));
}


// ====================================================
// Wrap selected functions from gdp/

// From gdp/gdp.h

// We need these specialized ..._print_stdout() wrappers because Node.js JS
// cannot supply a true stdout FILE* for calls to the node-ffi FFI.

// print a GCL (for debugging)
// extern void
// gdp_gcl_print(
//     const gdp_gcl_t *gclh,  // GCL handle to print
//     FILE *fp,               // file to print it to
//     int detail,             // not used at this time
//     int indent              // not used at this time
// );

// Forwards to gcp_gcl_print( const gdp_gcl_t *gclh, stdout, , , );
void
gdp_gcl_print_stdout(
    const gdp_gcl_t *gclh,  // GCL handle to print
    int detail,             // not used at this time
    int indent              // not used at this time
)
{
    gdp_gcl_print( gclh, stdout, detail, indent);
}


// print a message (for debugging)
// extern void
// gdp_datum_print(
//     const gdp_datum_t *datum, // message to print
//     FILE *fp                  // file to print it to
// );

// Forwards to gdp_datum_print( const gdp_datum_t *datum, stdout );
void
gdp_datum_print_stdout(
    const gdp_datum_t *datum // message to print
)
{
    gdp_datum_print( datum, stdout );
}


// ====================================================
// Wrap selected macros from gdp/gdp*.h in functions

// From gdp/gdp_stat.h

// #define GDP_STAT_NAK_NOTFOUND GDP_STAT_NEW(ERROR, GDP_COAP_NOTFOUND)

// Returns a EP_STAT with severity field ERROR; other fields see below.
EP_STAT
gdp_stat_nak_notfound()
{
	EP_STAT_32_64 rv;
    rv.as_EP_STAT  = (GDP_STAT_NEW(ERROR, GDP_COAP_NOTFOUND));
#if DEBUG
    fprintf(stderr, "gdp_stat_nak_notfound:"
                    "  Returning GDP_STAT_NAK_NOTFOUND = %x\n", rv.as_32_64_t);
    fflush(stderr);
#endif
    return (rv.as_EP_STAT);
}


// ====================================================
// Wrap selected macros from ep/ep*.h in functions

// From ep/ep_statcodes.h

// generic status codes
// #define EP_STAT_OK            EP_STAT_NEW(EP_STAT_SEV_OK, 0, 0, 0)

// Returns a EP_STAT with severity field OK; other fields are 0.
EP_STAT
ep_stat_ok()
{
	EP_STAT_32_64 rv;
    rv.as_EP_STAT = (EP_STAT_NEW(EP_STAT_SEV_OK, 0, 0, 0));
#if DEBUG
    fprintf(stderr, "ep_stat_ok:  Returning EP_STAT_OK = %x\n", rv.as_32_64_t);
    fflush(stderr);
#endif
    return (rv.as_EP_STAT);
}


// From ep/ep_statcodes.h

// common shared errors
// #define EP_STAT_END_OF_FILE   _EP_STAT_INTERNAL(WARN, EP_STAT_MOD_GENERIC, 3)

// Returns a EP_STAT with severity field WARN; other fields see below.
EP_STAT
ep_stat_end_of_file()
{
	EP_STAT_32_64 rv;
    rv.as_EP_STAT = ( _EP_STAT_INTERNAL(WARN, EP_STAT_MOD_GENERIC, 3) );
#if DEBUG
    fprintf(stderr, "ep_stat_end_of_file:"
                    "  Returning EP_STAT_END_OF_FILE = %x\n", rv.as_32_64_t);
    fflush(stderr);
#endif
    return (rv.as_EP_STAT);
}


// From ep/ep_stat.h

// predicates to query the status severity
// #define EP_STAT_ISOK(c)       (EP_STAT_SEVERITY(c) < EP_STAT_SEV_WARN)

// Return true iff the status severity field inside ep_stat is less
// than EP_STAT_SEV_WARN.
int /* Boolean */
ep_stat_isok(EP_STAT ep_stat) {
    int rv = EP_STAT_ISOK(ep_stat);
#if DEBUG
    fprintf(stderr, "ep_stat_isok: ep_stat = %x,",
	        ((EP_STAT_32_64) ep_stat).as_32_64_t );
    fprintf(stderr, "  Returning EP_STAT_ISOK() = %x\n", rv);
    fflush(stderr);
#endif
    return rv;
}


// From ep/ep_stat.h

// compare two status codes for equality
// #define EP_STAT_IS_SAME(a, b)   ((a).code == (b).code)

// Return true iff the status codes are the same
int /* Boolean */
ep_stat_is_same(EP_STAT a, EP_STAT b) {
    int rv = EP_STAT_IS_SAME(a, b);
#if DEBUG
    fprintf(stderr, "ep_stat_is_same: a = %x, b = %x",
	        (((EP_STAT_32_64) a).as_32_64_t ),
	        (((EP_STAT_32_64) b).as_32_64_t )
		   );
    fprintf(stderr, "  Returning EP_STAT_IS_SAME() = %x\n", rv);
    fflush(stderr);
#endif
    return rv;
}

File: lang/js/gdpjs/gdpjs_supt.js
/* vim: set ai sw=4 sts=4 ts=4 : */

// JavaScript support routines for GDP access from Node.js-based JavaScript
// 2014-11-02
// Alec Dara-Abrams



// JS wrappers for calls to libep, libgdp, and libgdpjs FFI functions =======

// Currently, we are using an eval-based mechanism to include the above
// libraries as well as this code; we hope to construct a Node.js module
// to better encapsulate all these support functions and variables.


/* Note:
   EP_STAT      JS<==>C  uint32
   gclHandle_t  JS<==>C  gdp_gcl_t *
   String       JS<==>C  gcl_name_t *
   TBD -- get these JS<-->C type correspondences documented here
   and in each ...._js() function below.
*/


/* EP_STAT */
function ep_stat_ok_js()
{
	return libgdpjs.ep_stat_ok();
}

/* C: size_t  <==> JS: Integer */
function sizeof_EP_STAT_in_bytes_js()
// Pull libgdp/libep information into JS.
// Really, a C compile-time constant we bring into JS
{
	return libgdpjs.sizeof_EP_STAT_in_bytes();
}

/* EP_STAT */
function gdp_init_js( /* String */ gdpd_addr )
{
	// libgdp.gdp_init() seems to have trouble passing in an explicit
	// null string but an undefined argument works OK to start the
	// library on the default gdpd host:port .
	// So we protect this call from incoming JS null strings.
	if ( gdpd_addr == "" )
	{	// DEBUG: console.log( 'then: gdpd_addr = \"' + gdpd_addr + '\"' );
		return libgdp.gdp_init( undefined );
	}
	else
	{	// DEBUG: console.log( 'else: gdpd_addr = \"' + gdpd_addr + '\"' );
		return libgdp.gdp_init( gdpd_addr );
	}
}

/* Boolean */
function ep_stat_isok_js( /* EP_STAT */ estat )
{
	return libgdpjs.ep_stat_isok(estat);
}

/* EP_STAT */
function gdp_stat_nak_notfound_js()
// return a constant "NAK NOTFOUND" EP_STAT value
{
	return libgdpjs.gdp_stat_nak_notfound();
}

/* EP_STAT */
function ep_stat_end_of_file_js()
// return a constant "END OF FILE" EP_STAT value
{
	return libgdpjs.ep_stat_end_of_file();
}

/* Boolean */
function ep_stat_is_same_js( estat_a, estat_b )
{
	return libgdpjs.ep_stat_is_same( estat_a, estat_b );
}

/* { error_code: EP_STAT, gclH: gclHandle_t }; */
function gdp_gcl_create_js( /* gcl_name_t */ gcliname,
                                             gclPtrPtr
						  )
{
	var gcl_Ptr;                                  // gclHandle_t 
	var gclPtrPtr = ref.alloc( gdp_gcl_tPtrPtr ); // gclHandle_t *
	var estat = libgdp.gdp_gcl_create(gcliname, gclPtrPtr);
	gcl_Ptr = gclPtrPtr.deref();
	return { error_code: estat, gclH: gcl_Ptr };
}

/* EP_STAT */
function gdp_gcl_parse_name_js( /* String */ xname, /* gcl_name_t */ gcliname )
{
	return libgdp.gdp_gcl_parse_name( xname, gcliname );
};

/* { error_code: EP_STAT, gclH: gclHandle_t }; */
function gdp_gcl_open_js( /* gcl_name_t */   gcliname,
                          /* gdp_iomode_t */ mode,
						                     gclPtrPtr
						)
{
	var gcl_Ptr;                                  // gclHandle_t 
	var gclPtrPtr = ref.alloc( gdp_gcl_tPtrPtr ); // gclHandle_t *
	var estat = libgdp.gdp_gcl_open(gcliname, mode, gclPtrPtr);
	gcl_Ptr = gclPtrPtr.deref();
	return { error_code: estat, gclH: gcl_Ptr };
}

/* void */
function ep_dbg_init_js( )
{
	libep.ep_dbg_init( );
};

/* String */
function ep_stat_tostr_js( /* EP_STAT */ estat,
                           /* String */  str,
						   /* Integer */ str_len
						 )
{
	return libep.ep_stat_tostr(estat, str, str.length);
};

/* void */
function ep_dbg_set_js( /* String */  ep_dbg_pattern )
{
	libep.ep_dbg_set( ep_dbg_pattern );
};

/* String */
function gdp_gcl_printable_name_js( /* gcl_name_t  */ gclname,
                                    /* gcl_pname_t */ gclpname )
{
	var rv_str;
	rv_str = libgdp.gdp_gcl_printable_name( gclname, gclpname );
	return rv_str;
}

/* void */
function gdp_gcl_print_stdout_js( gcl_Ptr, /* int */ detail, /* int */ indent )
{
	/* void */
    libgdpjs.gdp_gcl_print_stdout( gcl_Ptr, detail, indent );
}

/* String */
function gdp_get_pname_from_gclh_js( gcl_Ptr )
{
	var rv_str;
    rv_str = libgdpjs.gdp_get_pname_from_gclh( gcl_Ptr );
	return rv_str;
}

/* String */
function gdp_get_printable_name_from_gclh_js( gcl_Ptr )
{
	var rv_str;
	// Note, the below returns a pointer to a statically allocated gcl_pname_t,
	// currently, typedef char gcl_pname_t[GDP_GCL_PNAME_LEN + 1] in gdp.h
    rv_str = libgdpjs.gdp_get_printable_name_from_gclh( gcl_Ptr );
	return rv_str;
}

/* String */
function gdp_datum_getts_as_string_js( datum, /* Boolean */ human_format )
{
	var rv_str;
	// Note, the below returns a pointer to a statically allocated 
	// char tbuf[100] in gdpjs_supt.c/ep_time_as_string() .
    rv_str = libgdpjs.gdp_datum_getts_as_string( datum, human_format );
	// DEBUG console.log("gdp_datum_getts_as_string_js: rv_str = '" + rv_str + "'" );
	return rv_str;
}

/* datum */
function gdp_datum_new_js()
{
	return libgdp.gdp_datum_new();

};

/* EP_STAT */
function gdp_gcl_read_js( gclh, recno, datum )
{
	return libgdp.gdp_gcl_read( gclh, recno, datum );
}

/* gdp_recno_t */
function gdp_datum_getrecno_js( datum )
{
	return libgdp.gdp_datum_getrecno( datum );
}

/* EP_STAT */
function gdp_gcl_publish_buf_js( gcl_Ptr, datum, buf )
{
	var rv_estat;
	var temp_gdp_buf = libgdp.gdp_datum_getbuf( datum );
	libgdp.gdp_buf_write( temp_gdp_buf, buf, buf.length );
	rv_estat = libgdp.gdp_gcl_publish( gcl_Ptr, datum );
	return rv_estat;
};

/* EP_STAT */
function gdp_gcl_subscribe_no_timeout_no_callback_js( gclh, firstrec, numrecs )
// This is a limited-fucntionality version of
//   libgdp.gdp_gcl_subscribe(gclh, firstrec, numrecs, timeout, cbfunc, cbarg);
// We currently do not have a mechanism for invoking a JS callback function
// from down in the C level.  We are looking for a solution here using the
// Node.js FFI mechanism.
// A smaller issue is passing in the EP_TIME_SPEC timeout argument (a C struct).
// Node.js ref-struct should provide a straight-forward FFI implementation.
{
	var rv_estat;
	// EP_TIME_SPEC *timeout, gdp_gcl_sub_cbfunc_t cbfunc, void *cbarg 
	// are all null here for a simple subscribe variant.  See gdp/gdp-api.[ch]
	// gdp_gcl_subscribe() for details.
	rv_estat = libgdp.gdp_gcl_subscribe( gclh, firstrec, numrecs,
	                                     null, null, null );
	return rv_estat;
};

/* EP_STAT */
function gdp_gcl_multiread_no_callback_js( gclh, firstrec, numrecs )
// This is a limited-fucntionality version of
//   libgdp.gdp_gcl_multiread(gclh, firstrec, numrecs, cbfunc, cbarg);
// See the comments above for: gdp_gcl_subscribe_no_timeout_no_callback_js() .
{

	var rv_estat;
	// gdp_gcl_sub_cbfunc_t cbfunc, void *cbarg are both null here for a
	// simple multiread variant.  See gdp/gdp-api.[ch] gdp_gcl_multiread() .
	rv_estat = libgdp.gdp_gcl_multiread( gclh, firstrec, numrecs,
	                                     null, null);
	return rv_estat;
}

/* C: gdp_event_t *  <==>  node-ffi: gdp_event_tPtr  <==>  JS: Object */
function gdp_event_next_js( /* Boolean */ wait )
// Return value can be viewed by JS as an opaque handle for a gdp_event.
{
	var gev_Ptr;
	gev_Ptr = libgdp.gdp_event_next( wait );
	return gev_Ptr;
}

/* Integer */
function gdp_event_gettype_js( gev_Ptr )
// Arg gev_Ptr can be viewed by JS as an opaque handle for a gdp_event.
// Return value is an integer gdp event code -- see libgdp_h.js
{
	var evtype_int;
	evtype_int = libgdp.gdp_event_gettype( gev_Ptr );
	return evtype_int
}

/* C: gdp_datum_t *  <==>  JS: Object */
function gdp_event_getdatum_js( gev_Ptr )
// Arg gev_Ptr can be viewed by JS as an opaque handle for a gdp_event.
// Return value can be viewed by JS as an opaque handle for a gdp_datum.
{
	return libgdp.gdp_event_getdatum( gev_Ptr );
}

/* void */
function gdp_datum_print_stdout_js( datum )
// Arg datum can be viewed by JS as an opaque handle for a gdp_datum.
{
	/* void */
	libgdpjs.gdp_datum_print_stdout( datum );
}


/* EP_STAT */
function gdp_event_free_js( gev_Ptr )
// Arg gev_Ptr can be viewed by JS as an opaque handle for a gdp_event.
{
	return libgdp.gdp_event_free( gev_Ptr );
}

/* EP_STAT */
function gdp_gcl_close_js( gcl_Ptr )
{
	return libgdp.gdp_gcl_close(gcl_Ptr);
}

/* void */
function fflush_all_js()
// Explicit print buffer flush; sometimes Node.js may not empty buffers
{
	/* void */
	libgdpjs.fflush_all();
}

/* void */
function gdp_datum_free_js( datum )
{
	/* void */
	libgdp.gdp_datum_free( datum );
}



// Some GDP-related utility string manipulation functions ===================


/* Integer */
function String_to_Int( /* String */ str )
{
  if ( /^(\-|\+)?([0-9]+|Infinity)$/ .test(str) ) return Number(str);
  return NaN;
}


/* String */
function array_to_escaped_String( /* Array */ arry )
// Helpful for printing SHA-256 internal gcl names.
{
	var iToH = [ "0", "1", "2", "3", "4", "5", "6", "7",
	             "8", "9", "A", "B", "C", "D", "E", "F" ];
	var arry_as_string = "";
	for( var i = 0; i < arry.length; i++ )
	{
		var hiHexDigit = iToH[ Math.floor( arry[i] / 16 ) ];
		var loHexDigit = iToH[ arry[i] % 16 ];
		arry_as_string = arry_as_string + jsesc( String.fromCharCode( arry[i] ) );
		// console.log( "array_to_escaped_String: ", i, arry[i],
		//              hiHexDigit, loHexDigit, arry_as_string ); 
	}
	return arry_as_string;
}


/* String */
function array_to_String( /* Array */ arry )
// Helpful for printing SHA-256 internal gcl names.
{
	var iToH = [ "0", "1", "2", "3", "4", "5", "6", "7",
	             "8", "9", "A", "B", "C", "D", "E", "F" ];
	var arry_as_string = "";
	for( var i = 0; i < arry.length; i++ )
	{
		var hiHexDigit = iToH[ Math.floor( arry[i] / 16 ) ];
		var loHexDigit = iToH[ arry[i] % 16 ];
		arry_as_string = arry_as_string + String.fromCharCode( arry[i] );
		// console.log( "array_to_String: ", i, arry[i],
		//              hiHexDigit, loHexDigit, arry_as_string ); 
	}
	return arry_as_string;
}


/* String */
function array_to_HexDigitString( /* Array */ arry )
// Helpful for printing SHA-256 internal gcl names.
{
	var iToH = [ "0", "1", "2", "3", "4", "5", "6", "7",
	             "8", "9", "A", "B", "C", "D", "E", "F" ];
	var arry_as_hex = "";
	for( var i = 0; i < arry.length; i++ )
	{
		var hiHexDigit = iToH[ Math.floor( arry[i] / 16 ) ];
		var loHexDigit = iToH[ arry[i] % 16 ];
		arry_as_hex = arry_as_hex + hiHexDigit + loHexDigit;
		// console.log( "array_to_HexDigitString: ", i, arry[i],
		//              hiHexDigit, loHexDigit, arry_as_hex ); 
	}
	return arry_as_hex;
}


/* String */
function buffer_to_HexDigitString( /* Buffer */ bfr )
// Helper function for buffer_info_to_string() .
{
	var iToH = [ "0", "1", "2", "3", "4", "5", "6", "7",
	             "8", "9", "A", "B", "C", "D", "E", "F" ];
	var bfr_as_hex = "";
	for( var i = 0; i < bfr.length; i++ )
	{
		var hiHexDigit = iToH[ Math.floor( bfr[i] / 16 ) ];
		var loHexDigit = iToH[ bfr[i] % 16 ];
		bfr_as_hex = bfr_as_hex + hiHexDigit + loHexDigit;
		// console.log( "buffer_to_HexDigitString: ", i, bfr[i],
		//              hiHexDigit, loHexDigit, bfr_as_hex ); 
	}
	return bfr_as_hex;
}


/* String */
function buffer_info_to_string( /* string */ bfrName, /* Buffer */ bfr )
// Used for debugging node-ref "types" which may be implemented as a JS Buffer.
{
	var isNonNullObject = ( bfr != null ) && ( typeof(bfr) == 'object' );
	var str = "";
	// console.log( "buffer_info_to_string: " );
	// console.log( "isNonNullObject =\"" + isNonNullObject + "\"" );
	// console.log( "bfr =\"" + bfr + "\"" );
	// console.log( "typeof(bfr) =\"" + typeof(bfr) + "\"" );
	//if ( typeof(bfr) == 'object' && bfr.constructor.name == "Buffer" )
	if ( isNonNullObject )
	{
		if ( bfr.constructor.name == "Buffer" )
		{
			str = str + bfrName + " =\"" + bfr + "\"\n";
			str = str + "typeof(" + bfrName + ") =\"" + typeof(bfr) + "\"\n";
			str = str + bfrName + ".constructor.name =\"" + 
			            bfr.constructor.name + "\"\n";
			str = str + bfrName + ".length =\"" + bfr.length + "\"\n";
			str = str + "buffer_to_HexDigitString(" + bfrName + ") =\"" +
			            buffer_to_HexDigitString( bfr ) + "\"\n";
			str = str + bfrName + ".toString() =\"" + bfr.toString() + "\"\n";
			str = str + "ref(" + bfrName + ").getType =\"" +
			            ref.getType(bfr) + "\"\n";
			str = str + bfrName + ".deref =\"" + bfr.deref() + "\"";
		} else
		{
			str = "\"" + bfrName + "\" is not a Buffer object, it is a \"" +
				  bfr.constructor.name + "\"";
		}
	} else
	{
		str = "\"" + bfrName +
		      "\" is not a non-null Buffer object, typeof = \"" +
			  typeof( bfr ) + "\", value = \"" + bfr + "\"" ;
	}
	return str;
} /* end function buffer_info_to_string() */


/* String */
function get_datum_buf_as_string( datum )
// Return the bytes of a GDP datum as a JS String (which is able to
// contain nulls).
{
	var datum_dlen = libgdp.gdp_datum_getdlen( datum );

// DEBUG - keep this output here until we've looked this code over more.
DEBUG_get_datum_buf_as_string = false;

if ( DEBUG_get_datum_buf_as_string ) {
	console.log( "get_datum_buf_as_string: After call to gdp_datum_print() gcp_datum_getdlen() = \"" + datum_dlen + "\"" );
}
	var gdp_buf = libgdp.gdp_datum_getbuf( datum );
if ( DEBUG_get_datum_buf_as_string ) {
	console.log( "get_datum_buf_as_string: After call to gdp_datum_getbuf() gcp_datum_getbuf() = \"" + gdp_buf + "\"" );
}
	// curItem will hold the contents of the GCL item to be read here
	var curItem   = new buf_t(1000); // TBD check size??
	var gdp_buf_size = libgdp.gdp_buf_read( gdp_buf, curItem, curItem.length );
if ( DEBUG_get_datum_buf_as_string ) {
	console.log( "get_datum_buf_as_string: After call to gdp_buf_read() gdp_buf_read() = \"" + gdp_buf_size + "\"" );
	console.log( "get_datum_buf_as_string: After call to gdp_buf_read() curItem = \"" + curItem.toString() + "\"\n" );
}
	var rvString = '';
	for ( var i = 0; i < gdp_buf_size; i++ )
	{  rvString = rvString + String.fromCharCode( curItem[i] );
	}
if ( DEBUG_get_datum_buf_as_string ) {
	console.log("get_datum_buf_as_string: rvString = '" + rvString + "'" );
}
	return rvString;
} /* function get_datum_buf_as_string() */

